MTM-55444: edge-databroker: add new uninstall steps file (#1443)
SoftwareAG/c8y-docs: content/edge/databroker-bundle/databroker-edge-uninstall.md

---
weight: 50
title: Uninstalling the Messaging Service and the microservice-based data broker
layout: redirect
---

The bundle for the Messaging Service and data broker microservice does not have an uninstall command. 
If a mistake is made or an installation fails, the bundle components must be uninstalled manually.
Afterwards, the installation can be reattempted.

### To uninstall manually
1. Login to {{< product-c8y-iot >}} Edge using SSH.
   
2. Uninstall the Pulsar Helm chart.
```shell
sudo helm uninstall pulsar -n c8y-messaging-service
```
3. Wait for the pulsar pods to terminate.  
   This command updates its state every two seconds. When it reports that there are no resources, press Ctrl+C to return to the command line.

```shell

> watch sudo kubectl get pods -n c8y-messaging-service

Reviewer: Suggestion: Might need to double-check that it works in the Edge environment (it should) but `sudo watch ...` is probably more polite than `watch sudo ...` because it only has to run `sudo` once.
```suggestion
sudo watch kubectl get pods -n c8y-messaging-service
```

Author: No it doesn't work (I've tried as I favor and always try one sudo command over repeated). The sudo permissions have restrictions/limitations on edge VMs image I've used

Reviewer: OK, no worries, it was worth a try but I understand the security concerns that Edge has, makes sense they would restrict it.

```
4. Delete Pulsar's persistent volume claims.
```shell
sudo kubectl delete pvc -n c8y-messaging-service --all
```
5. Find the names of the persistent volumes used by the Pulsar components.
   

>    Here the GREP command is used to filter the output. The names are in the output's leftmost column.

Reviewer: Suggestion: I know "grep" is an acronym but in practice it's never written in uppercase.  I would go with just plain grep, or `grep` - but if you choose the latter there are other commands that would need to be presented the same way (I spotted jq and cURL, there maybe others)
```suggestion
   Here the grep command is used to filter the output. The names are in the output's leftmost column.
```

Author: I think cURL is lower (or natural) case, so changing GREP and JQ to lower (leaving cURL as it is) is fine. I'll do so

```shell
sudo kubectl get pv | grep pulsar
journal0     2Gi        RWO            Retain           Released    c8y-messaging-service/pulsar-bookie-journal-pulsar-bookie-0      local-storage            3h
ledgers0     10Gi       RWO            Retain           Released    c8y-messaging-service/pulsar-bookie-ledgers-pulsar-bookie-0      local-storage            3h
zookeeper0   2Gi        RWO            Retain           Released    c8y-messaging-service/pulsar-zookeeper-data-pulsar-zookeeper-0   local-storage            3h
```
6. Delete each of the persistent volumes found in the previous step.
```shell 
sudo kubectl delete pv journal0 ledgers0 zookeeper0
```
7. Delete the directories used by the persistent volumes.
```shell
sudo rm -rf /opt/bookie/ledgers /opt/bookie/journal /opt/zookeeper
```

> 8. Unsubscribe the edge tenant from and delete, the databroker-agent-server microservice.

Reviewer: Questions:
1. Shouldn't this be done first, before uninstalling the messaging components that the broker relies upon?
2. Can't this be done much more easily from the UI if you have access to that?  I do understand that we might need to present both methods.

Author: It works either way (it works the way I've described as that what I did to test - I haven't tried the other way but I'm sure it'd work also). I might not have time to retest the other way round as I'm on leave from tomorrow - so might be safer to leave as it is just in case there are issues

Author: @sagscmi please confirm this is okay (on leave from end of today) so I can merge

Reviewer: No objection from my side.

   The databroker-agent-server application ID is required for the HTTP requests used to do both of these.   

>    To find its application ID, you can use a cURL command of the following form and redirect the output to a JQ command to make it easier to read the JSON response.

Reviewer: ```suggestion
   To find its application ID, you can use a cURL command of the following form and redirect the output to a jq command to make it easier to read the JSON response.
```

   The general structure of the command is:
```shell

> curl http://<EDGE-HOSTNAME>/application/applicationsByName/databroker-agent-server -su <TENANT>/admin | jq --indent 2

Reviewer 2: <Tenant>/admin could be changed to something of the form <Tenant>/<Tenant username> to accommodate the changes which allow the user to select the tenant username rather than use 'admin'. This also applies to the stages below:
To unsubscribe, use a command of the following form
To delete, use a command of the following form

Author: @TzviLipschitz, thanks. Now updated to use placeholder

---
weight: 50

> title: Uninstalling the Messaging Service and the microservice-based data broker on Edge

Reviewer 3: ```suggestion
title: Uninstalling the Messaging Service and the microservice-based data broker
```

layout: redirect
---

The bundle for the Messaging Service and data broker microservice does not have an uninstall command. 

> If a mistake is made, or and installation fails, the bundle components must be manually uninstalled.

Reviewer 3: ```suggestion
If a mistake is made or an installation fails, the bundle components must be uninstalled manually.
```

> Once manually uninstalled, the installation can be reattempted.

Reviewer 3: ```suggestion
Afterwards, the installation can be reattempted.
```

> #### To manually uninstall:

Reviewer 3: ```suggestion
### To uninstall manually
```

1. Login to {{< product-c8y-iot >}} Edge using SSH.
   

> 2. Uninstall the Pulsar helm chart.

Reviewer 3: ```suggestion
2. Uninstall the Pulsar Helm chart.
```

```shell
sudo helm uninstall pulsar -n c8y-messaging-service
```
3. Wait for the pulsar pods to terminate.  

>    This command updates its state every 2 seconds. When it reports there are no resources, press CTL-C to return to the command line.

Reviewer 3: ```suggestion
   This command updates its state every two seconds. When it reports that there are no resources, press Ctrl+C to return to the command line.
```

```shell
watch sudo kubectl get pods -n c8y-messaging-service
```
4. Delete Pulsar's persistent volume claims.
```shell
sudo kubectl delete pvc -n c8y-messaging-service --all
```
5. Find the names of the persistent volumes used by the Pulsar components.
   
   Here the GREP command is used to filter the output. The names are in the output's leftmost column.

```shell
sudo kubectl get pv | grep pulsar
journal0     2Gi        RWO            Retain           Released    c8y-messaging-service/pulsar-bookie-journal-pulsar-bookie-0      local-storage            3h
ledgers0     10Gi       RWO            Retain           Released    c8y-messaging-service/pulsar-bookie-ledgers-pulsar-bookie-0      local-storage            3h
zookeeper0   2Gi        RWO            Retain           Released    c8y-messaging-service/pulsar-zookeeper-data-pulsar-zookeeper-0   local-storage            3h
```
6. Delete each of the persistent volumes found in step 5.
```shell 
sudo kubectl delete pv journal0 ledgers0 zookeeper0
```
7. Delete the directories used by the persistent volumes.
```shell
sudo rm -rf /opt/bookie/ledgers /opt/bookie/journal /opt/zookeeper
```

> 8. Unsubscribe the edge tenant from and undeploy, the databroker-agent-server microservice.

Reviewer 4: In fact unsubscribe==undeploy so maybe something like `unsubscribe and delete data broker ms from the edge tenant`?

Author: Done - it matches language used later also

   The databroker-agent-server application ID is required for the HTTP requests used to do both of these.   

>    To find its application ID, you can use a CURL command of the following form and redirect the output to a JQ command to make it easier to read the JSON response.

Reviewer 3: ```suggestion
   To find its application ID, you can use a cURL command of the following form and redirect the output to a JQ command to make it easier to read the JSON response.
```

   In the following example, it has the value '19' (the penultimate JSON field).
```shell
curl http://myown.iot.com/application/applicationsByName/databroker-agent-server -su edge/admin | jq --indent 2
Enter host password for user 'edge/admin':
{
  "next": "https://myown.iot.com/application/applications?pageSize=5&currentPage=2",
  "self": "https://myown.iot.com/application/applications?pageSize=5&currentPage=1",
  "statistics": {
    "pageSize": 5,
    "currentPage": 1
  },
  "applications": [
    {
      "owner": {
        "self": "https://myown.iot.com/tenant/tenants/edge",
        "tenant": {
          "id": "edge"
        }
      },
      "requiredRoles": [
        "ROLE_DATA_BROKER_READ",
        "ROLE_INVENTORY_READ",
        "ROLE_INVENTORY_CREATE",
        "ROLE_ALARM_ADMIN"
      ],
      "manifest": {
        "livenessProbe": {
          "failureThreshold": 3,
          "periodSeconds": 10,
          "timeoutSeconds": 5,
          "successThreshold": 1,
          "initialDelaySeconds": 50,
          "httpGet": {
            "path": "/health",
            "port": 80
          }
        },
        "requiredRoles": [

>           "ROLE_DATA_BROKER_READ",

Reviewer 4: Maybe remove unnecessary payload data and just put `...` leaving just a basic fields like type, self, id, key, availability and so on.

Author: Good idea - it will help reader focus on area of interest

Author: Done

    }
  ]
}
```
   Use that ID value in the request to unsubscribe from the microserive, and again when deleting it. 

   To unsubscribe, use a command of the following form:
```shell
curl -X DELETE http://<EDGE-HOSTNAME>/tenant/tenants/<TENANT>/applications/<APP-ID> -u <TENANT>/admin -v

```
   This should return an HTTP 204 (no content) response. For example: 
```shell
curl -X DELETE http://myown.iot.com/tenant/tenants/edge/applications/19 -u edge/admin -v
Enter host password for user 'edge/admin':
* About to connect() to myown.iot.com port 80 (#0)
*   Trying 127.0.0.1...
* Connected to myown.iot.com (127.0.0.1) port 80 (#0)
* Server auth using Basic with user 'edge/admin'
> DELETE /tenant/tenants/edge/applications/19 HTTP/1.1
> Authorization: Basic ZWRnZS9hZG1pbjpFZGdlN184OV9fX19fX18=
> User-Agent: curl/7.29.0
> Host: myown.iot.com
> Accept: */*
>
< HTTP/1.1 204 No Content
< Date: Thu, 07 Sep 2023 16:27:32 GMT
< Connection: keep-alive
< X-Content-Type-Options: nosniff
< X-XSS-Protection: 1; mode=block
< Cache-Control: no-cache, no-store, max-age=0, must-revalidate
< Pragma: no-cache
< Expires: 0
< X-Frame-Options: DENY
<
* Connection #0 to host myown.iot.com left intact
```

>    To delete, the use a command of the form:

Reviewer 3: ```suggestion
   To delete, use a command of the following form:
```

```shell
curl -X DELETE http://<EDGE-HOSTNAME>/application/applications/<APP-ID> -su <TENANT>/admin -v
```   

>    This should return an HTTP 204 (no content) response. For example:

Reviewer 3: ```suggestion
   This returns an HTTP 204 (no content) response. For example:
```
